### PR TITLE
really fix ensure_puppetclasses

### PR DESCRIPTION
**Pull request by evgeni**

### CODE REVIEW

File: plugins/module_utils/foreman_helper.py [hunk 1]
```diff
@@ -1096,14 +1096,14 @@ def _update_entity(self, resource, desired_entity, current_entity, params, forem
...
-            payload['id'] = current_entity['id']
+            payload['id'] = current_flat_entity['id']
             if not self.check_mode:
                 if params:
                     payload.update(params)
                 return self.resource_action(resource, 'update', payload)
             else:
                 # In check_mode we emulate the server updating the entity
-                fake_entity = current_entity.copy()
+                fake_entity = current_flat_entity.copy()
```
Comment (evgeni) [inline plugins/module_utils/foreman_helper.py:1106]: this one is "interesting", as in theory we need the `current_entity`, but then I am not sure that `update(payload)` would actually do something useful :(

**Comment by evgeni** [comment]
Heh, so @mdellweg, this now fails the fixtures for `host`, because :drum: you fixed the "update host puppetclasses" branch, but it still wasn't just totally right :laughing:

File: tests/test_playbooks/fixtures/host-0.yml [hunk 2]
```diff
@@ -33,7 +33,7 @@ interactions:
       Foreman_current_organization:
       - ; ANY
       Foreman_version:
-      - 2.3.0
+      - 3.0.0-rc1
```
Comment (evgeni) [inline tests/test_playbooks/fixtures/host-0.yml]: so technically this is not run against a plain foreman, but foreman+foreman_puppet now… let's see if old apidocs still match :D

File: tests/test_playbooks/fixtures/host-26.yml [hunk 3]
```diff
@@ -114,6 +114,66 @@ interactions:
...
+      Accept:
+      - application/json;version=2
+      Accept-Encoding:
+      - gzip, deflate
+      Connection:
+      - keep-alive
+      User-Agent:
+      - apypie (https://github.com/Apipie/apypie)
+    method: GET
+    uri: https://foreman.example.org/foreman_puppet/api/environments?search=name%3D%22production%22&per_page=4294967296
```
Comment (evgeni) [inline tests/test_playbooks/fixtures/host-26.yml]: meh, this won't work, I bet

**Comment by evgeni** [comment]
this is missing:
- [x] changelog
- [x] test for the actual issue in #1231, setting `description` (or really, any other normal attribute) while also providing (but not changing) `puppetclasses`

File: plugins/module_utils/foreman_helper.py (hunk 1 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:1106]: huh, how about

```
fake_entity = current_entity.copy()
fake_entity.update(current_flat_entity)
fake_entity.update(payload)
```

Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:1106]: @mdellweg what did we end up deciding here? I cant' remember :(

**Comment by evgeni** [comment]
CI errors unrelated

File: plugins/module_utils/foreman_helper.py (hunk 1 above)
Comment (mdellweg) [reply plugins/module_utils/foreman_helper.py:1106]: As the payload is only updating the flattened values, copying the `current_entity` is probably dangerous / misleading.
So i think, we should leave it as is.

File: plugins/module_utils/foreman_helper.py [hunk 4]
```diff
@@ -1779,10 +1779,11 @@ def build_fqn(name, parent=None):
 
 # Helper for puppetclasses
 def ensure_puppetclasses(module, entity_type, entity, expected_puppetclasses=None):
-    if expected_puppetclasses:
+    if expected_puppetclasses is not None:
         puppetclasses_resource = '{0}_classes'.format(entity_type)
         expected_puppetclasses = module.find_puppetclasses(expected_puppetclasses, environment=entity['environment_id'], thin=True)
-        current_puppetclass_ids = entity.get('puppetclass_ids', [])
+        current_puppetclasses = entity.get('puppetclasses', [])
```
Comment (mdellweg) [inline plugins/module_utils/foreman_helper.py:1785]: Will this be available in the `fake_entity` created from the `flattened_current_entity`?

**Review by mdellweg — APPROVED** [review]
If i understand the change correctly, we may report slightly different (but still correct on their own) results in "normal" and "check mode" operation. However, this may not even be new.
This solves a bug; we should just take it.

File: plugins/module_utils/foreman_helper.py (hunk 1 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:1106]: OK

**Comment by evgeni** [comment]
> If i understand the change correctly, we may report slightly different (but still correct on their own) results in "normal" and "check mode" operation. However, this may not even be new.

Yeah, the check mode output was always a bit opportunistic, as we never ask the server how it would change it.

File: plugins/module_utils/foreman_helper.py (hunk 4 above)
Comment (evgeni) [reply plugins/module_utils/foreman_helper.py:1785]: No.

**Pull request merged**